### PR TITLE
Adds new `match()` method

### PR DESCRIPTION
**Pull request by mertasan**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes

This is an alternative to #406.

```php
it('database test', function () {
    
    expect(new Foo())
        ->foo->toBe('foo')
        ->match(DB::getDriverName(), [
            'mysql'  => fn ($foo) => $foo->mysql->toBeTrue(),
            'sqlite' => fn ($foo) => $foo->sqlite->toBeTrue(),
        ])
        ->bar->toBe('bar');

});
```

```php
it('builder test', function ($driver) {
    
    expect(new Builder())
        ->match($driver, [
            'mysql'   => fn ($b) => $b->where('...')->toBe(...),
            'postgre' => fn ($b) => $b->whereRaw('...')->toBe(...),
        ]);
    
})->with([
    'mysql',
    'postgre',
]);
```

For more examples, please review the tests.

Thank you.

### CODE REVIEW

**Comment by nunomaduro** [comment]
@lukeraymonddowning What do you think?

File: src/Expectation.php [hunk 1]
```diff
@@ -177,6 +177,56 @@ public function sequence(...$callbacks): Expectation
...
+                continue;
+            }
+
+            (new self($this->value))->toEqual($callback);
+
+            break;
+        }
+
+        if (!$matched) {
+            test()->addWarning('No item found matching "' . $subject . '".');
```
Comment (nunomaduro) [inline src/Expectation.php:224]: This should be removed.

File: src/Expectation.php [hunk 2]
```diff
@@ -177,6 +177,56 @@ public function sequence(...$callbacks): Expectation
...
+            }
+
+            $matched = true;
+
+            if (is_callable($callback)) {
+                $callback(new self($this->value));
+                continue;
+            }
+
+            (new self($this->value))->toEqual($callback);
```
Comment (nunomaduro) [inline src/Expectation.php:218]: `this->and`

File: src/Expectation.php (hunk 1 above)
Comment (nunomaduro) [reply src/Expectation.php:224]: And add a `$this->toHaveKey(` before all this code.

Comment (mertasan) [reply src/Expectation.php]: @nunomaduro Can you explain this in more detail please? I don't understand where to add it. `$this->toHaveKey(..?.)`

**Comment by lukeraymonddowning** [comment]
I actually like this a lot. It's a sharp knife, for sure, but a valuable one I think.

**Pull request merged**